jni/libucommon/sources/utils/zerofill.1
.\" zerofill - clear unused disk space and blank devices.
.\" Copyright (c) 2010 David Sugar <dyfet@gnutelephony.org>
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc.,59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH zerofill "1" "January 2010" "GNU uCommon" "GNU Telephony"
.SH NAME
zerofill \- clear unused disk space and blank devices.
.SH SYNOPSIS
.B zerofill
.RI [ options ]
.RI [ devices ]
.I ...
.br
.SH DESCRIPTION
This command is a relative of nullfile.  Like nullfile, it can be used
to write all unused disk space with '0', thereby making file systems that
are stored on virtual machines easier to pack.  In addition, zerofill can
fill unused disk space with random data before zeroing, which can be useful
to remove forensic trace of deleted files.  Finally, zerofile can be told
to operate on and zero (erase) complete disk devices.
.SH OPTIONS
.TP
.BI \-\-random= count
Number of passes to use writing random data before wiping data.
.TP
.B \-\-help
Outputs help screen for the user.
.SH AUTHOR
.B zerofill
was written by David Sugar <dyfet@gnutelephony.org>.
.SH "REPORTING BUGS"
Report bugs to bug-commoncpp@gnu.org.
.SH COPYRIGHT
Copyright \(co 2010 David Sugar, Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.




jni/libucommon/sources/utils/zerofill.cpp
// Copyright (C) 2010 David Sugar, Tycho Softworks.
//
// This file is part of GNU uCommon C++.
//
// GNU uCommon C++ is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// GNU uCommon C++ is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with GNU uCommon C++.  If not, see <http://www.gnu.org/licenses/>.

#include <ucommon/secure.h>

using namespace UCOMMON_NAMESPACE;

static shell::flagopt helpflag('h',"--help",    _TEXT("display this list"));
static shell::flagopt althelp('?', NULL, NULL);
static shell::numericopt passes('r', "--random", _TEXT("optional passes with randomized data (0-x)"), "count", 0);

static unsigned char buffer[65536];
static bool live = false;
static bool temp = false;

static void cleanup(void)
{
    if(temp)
        fsys::erase("zerofill.tmp");
    if(live)
        shell::printf("\n");
    live = false;
}

static void zerofill(void)
{
    fsys::offset_t pos = 0l;
    ssize_t size = 65536;
    unsigned long count = 0;
    fsys_t fs;

    temp = true;

    fs.open("zerofill.tmp", 0666, fsys::STREAM);
    if(!is(fs))
        shell::errexit(1, "*** zerofill: %s\n",
            _TEXT("cannot create temporary file"));

    live = true;
    while(size == 65536 && live) {
        unsigned pass = 0;

        while(pass < (unsigned)*passes) {
            fs.seek(pos);
            Random::fill(buffer, sizeof(buffer));
            fs.write(buffer, sizeof(buffer));
            ++pass;
        }

        fs.seek(pos);
        memset(buffer, 0, sizeof(buffer));
        size = fs.write(buffer, sizeof(buffer));
        pos += sizeof(buffer);
        if(++count >= 256) {
            shell::printf(".");
            count = 0;
        }
    }

    cleanup();

    if(fs.err() != ENOSPC)
        shell::errexit(3, "*** zerofill: %s\n",
            _TEXT("failed before end of space"));
}

static void zerofill(const char *devname)
{
    fsys::fileinfo_t ino;
    fsys::offset_t pos = 0l;
    ssize_t size = 65536;
    unsigned long count = 0;
    fsys_t fs;

    if(fsys::info(devname, &ino))
        shell::errexit(5, "*** zerofill: %s: %s\n",
            devname, _TEXT("cannot access"));

    if(!fsys::is_disk(&ino))
        shell::errexit(6, "*** zerofill: %s: %s\n",
            devname, _TEXT("not block device"));

    fs.open(devname, fsys::WRONLY);
    if(fs.err())
        shell::errexit(5, "*** zerofill: %s: %s\n",
            devname, _TEXT("cannot modify"));

    shell::printf("%s", devname);

    live = true;
    while(size == 65536 && live) {
        unsigned pass = 0;

        while(pass < (unsigned)*passes) {
            fs.seek(pos);
            Random::fill(buffer, sizeof(buffer));
            fs.write(buffer, sizeof(buffer));
            ++pass;
        }

        fs.seek(pos);
        memset(buffer, 0, sizeof(buffer));
        size = fs.write(buffer, sizeof(buffer));
        pos += sizeof(buffer);
        if(++count >= 256) {
            shell::printf(".");
            count = 0;
        }
    }

    cleanup();

    if(fs.err() != ENOSPC)
        shell::errexit(3, "*** zerofill: %s\n",
            _TEXT("failed before end of space"));
}

PROGRAM_MAIN(argc, argv)
{
    shell::bind("zerofill");
    shell args(argc, argv);
    unsigned count = 0;

    if(*passes < 0)
        shell::errexit(2, "*** zerofill: random: %ld: %s\n",
            *passes, _TEXT("negative random passes invalid"));

    if(is(helpflag) || is(althelp)) {
        printf("%s\n", _TEXT("Usage: zerofill [options] path..."));
        printf("%s\n\n", _TEXT("Erase and fill unused space with zeros"));
        printf("%s\n", _TEXT("Options:"));
        shell::help();
        printf("\n%s\n", _TEXT("Report bugs to dyfet@gnu.org"));
        PROGRAM_EXIT(0);
    }

    secure::init();

    shell::exiting(&cleanup);

    if(!args())
        zerofill();

    while(count < args() && live)
        zerofill(args[count++]);

    PROGRAM_EXIT(0);
}

